src/variables.jl

```diff
@@ -33,3 +33,26 @@ A [`AddVariableNotAllowed`](@ref) error is thrown if adding variables cannot be
 done in the current state of the model `model`.
 """
 add_variable(model::ModelLike) = throw(AddVariableNotAllowed())
+
+#function supports_constrained_variables(
+#    model::ModelLike, S::Type{<:AbstractVectorSet})
+#    return MOI.supports_constraint(model, MOI.VectorOfVariables, S)
+#end
+function add_constrained_variable(model::ModelLike, set::AbstractScalarSet)
```

Comment:
How does one add a variable with multiple sets? Like `LessThan`, `GreaterThan`, and `ZeroOne`? How does JuMP know when to call `add_constrained_variable`, or `add_variable` and then `add_constraint`?

Reply:
I would suggest the following:
```julia
@variable(model, x[1:n] in set)
```
does `MOI.add_constrained_variables(model, set)`.
If there are multiple set, you can use `add_constrained_variables` for one of the sets and `add_constraint` for the other ones.
There are cases where there is an impact on the formulation depending on which one you use in `add_constrained_variables` but the only cases impacted are cases which were worse before variable bridged and `add_constrained_variables`.
* If you use `add_constrained_variables` and it bridges the variable, then `add_constraint` on the bridged variable cannot work since it is an affine expression for the solver, not just a variable. It is then *force-bridged* by `FunctionizeBridge` into affine function-in-set. So it is better to do `add_constrained_variables` with a set `S` such that `supports_constraint(MOI.VectorOfVariables, S)` is true.
* For Mosek, SDP variables can only be added with `add_constrained_variables` so it should have priority.
For
```julia
@variable(model, -1 <= x[1:n] <= 1, Int)
```
it does not really matter for which set we use `add_constrained_variables` if the LP solver does not use variable bridges.

Reply:
> For Mosek, SDP variables can only be added with add_constrained_variables so it should have priority.

How does JuMP determine this priority? How can we add a PSD matrix that is constrained to be integer?

Reply:
You cannot do that with Mosek. If the user tries to create VectorOfVariables-in-PSD with `add_constraint` it currently work because of a hack in the wrapper, i.e. I delete the variables created and add PSD ones but I am tempted to remove that once `add_constrained_variables` exists.
If the user tried to set PSD variables as binary, they will get a useful error from Mosek saying that PSD variables are handled differently and can't be binary.

Reply:
> You cannot do that with Mosek.

Sure, but what about a different solver in future.

Reply:
> How does JuMP determine this priority?

I would say that in
```julia
@variable(model, x[1:3] in set, binary = true)
```
you do `MOI.add_constrained_variables(model, set)` and then you set the `binary` with `MOI.add_constraint` because the `set` variable-constraint is more likely to be bridged than the `MOI.ZeroOne` variable-constraint.
Another approach is to check with `MOI.supports_constraint` which variable-constraint is supported and use `MOI.add_constrained_variables` for one that is not supported so that it can be variable-bridged.